sql/src/main/java/io/crate/executor/transport/ShardUpsertRequest.java

```diff
@@ -0,0 +1,323 @@
+/*
+ * Licensed to CRATE Technology GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.executor.transport;
+
+import com.carrotsearch.hppc.IntArrayList;
+import com.google.common.collect.Iterators;
+import io.crate.Constants;
+import io.crate.Streamer;
+import io.crate.planner.symbol.Reference;
+import io.crate.planner.symbol.Symbol;
+import org.elasticsearch.action.support.replication.ShardReplicationOperationRequest;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.lucene.uid.Versions;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class ShardUpsertRequest extends ShardReplicationOperationRequest<ShardUpsertRequest> implements Iterable<ShardUpsertRequest.Item> {
+
+    /**
+     * A single update item.
+     */
+    public static class Item implements Streamable {
+
+        private String id;
+        private String routing;
+        private Symbol[] assignments;
+        private long version = Versions.MATCH_ANY;
+        @Nullable
+        private Object[] missingAssignments;
+        @Nullable
+        private Streamer[] streamers;
```

Comment:
maybe be a bit more verbose here and name it missingAssignmentStreamers to indicate that they're just used to stream missingAssigments?